apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/util/CustomizeExpression.java

```diff
@@ -37,7 +37,7 @@ public class CustomizeExpression {
     private static final ILog LOGGER = LogManager.getLogger(CustomizeExpression.class);
 
     public static Map<String, Object> evaluationContext(Object[] allArguments) {
-        Map<String, Object> context = new HashMap<>();
+        Map<String, Object> context = new HashMap<>(allArguments.length, 1);
```

Comment:
Why change the load factor to `1`?

Reply:
Using the default threshold, it will still expand when capacity is `0.75 * size()`, and I want to prevent it from expanding completely

Reply:
I think this can't avoid that case actually. Because factor is not determining this.

Reply:
I'm a little confused

>  if (++size > threshold)
            resize();

It is true that it is above the calculated threshold for expansion

If I'm wrong can you explain exactly what it is that determines expansion

Reply:
If you think the load factor is critical, you may provide the report of microbenchmarks.

Reply:
> 
    @Test
    public void expandTest(){
        HashMap<String, String> map = new HashMap<>(20);
        final long start1 = System.nanoTime();
        for (int i = 0; i < 20; i++) {
            map.put("key" + i, "value");
        }
        System.out.println("Expansion takes nanoTime = " + (System.nanoTime() - start1));

        HashMap<String,String> map2 = new HashMap<>(20, 1);
        final long start2 = System.nanoTime();
        for (int i = 0; i < 20; i++) {
            map2.put("key" + i, "value");
        }
        System.out.println("Without expansion takes nanoTime = " + (System.nanoTime() - start2));
    }

> Expansion takes nanoTime = 31689
Without expansion takes nanoTime = 7528

Reply:
Is this performance sensitive? Bigger factor actually costs more memory. Generally, this is a tradeoff.

Reply:
I think performance should take precedence over memory consumption for `agent`

Do you think?

Reply:
> > 
> 
> ```
> @Test
> public void expandTest(){
>     HashMap<String, String> map = new HashMap<>(20);
>     final long start1 = System.nanoTime();
>     for (int i = 0; i < 20; i++) {
>         map.put("key" + i, "value");
>     }
>     System.out.println("Expansion takes nanoTime = " + (System.nanoTime() - start1));
> 
>     HashMap<String,String> map2 = new HashMap<>(20, 1);
>     final long start2 = System.nanoTime();
>     for (int i = 0; i < 20; i++) {
>         map2.put("key" + i, "value");
>     }
>     System.out.println("Without expansion takes nanoTime = " + (System.nanoTime() - start2));
> }
> ```
> 
> > Expansion takes nanoTime = 31689
> > Without expansion takes nanoTime = 7528

I deeply doubt this test. This can hardly be called a microbenchmark.

If you swap two cases, the result will be the opposite!

```java
HashMap<String,String> map2 = new HashMap<>(20, 1);
final long start2 = System.nanoTime();
for (int i = 0; i < 20; i++) {
    map2.put("key" + i, "value");
}
System.out.println("Without expansion takes nanoTime = " + (System.nanoTime() - start2));

HashMap<String, String> map = new HashMap<>(20);
final long start1 = System.nanoTime();
for (int i = 0; i < 20; i++) {
    map.put("key" + i, "value");
}
System.out.println("Expansion takes nanoTime = " + (System.nanoTime() - start1));
```

this just generated (with extremely large fluctuation)

```
Without expansion takes nanoTime = 181705
Expansion takes nanoTime = 90335
```

Reply:
> Is this performance sensitive? Bigger factor actually costs more memory. Generally, this is a tradeoff.

I wrote a simple microbenchmark to test the impact of the load factor, which can be found here,

https://gist.github.com/lujiajing1126/17c5372af0e940a21067a3e6f8554594

I scan the parameter space from 1 to 20 (not even), and the result shows no significant difference.

So I'm against this kind of modification.

Reply:
@Oliverwqcwrw your test is not measured correctly due to the JVM warmup. I also think this load factor should not be tempered here unless there's solid empirical/ mathematical evidence.

Reply:
> @Oliverwqcwrw your test is not measured correctly due to the JVM warmup. The first part will always take much longer unless you run few things before. I also think this load factor should not be tempered here unless there's solid empirical/ mathematical evidence.

Thanks for your solution, I do wonder why my test result is so strange, it seems that the future test will write a benchmark test

Reply:
> > Is this performance sensitive? Bigger factor actually costs more memory. Generally, this is a tradeoff.
> 
> I wrote a simple microbenchmark to test the impact of the load factor, which can be found here,
> 
> https://gist.github.com/lujiajing1126/17c5372af0e940a21067a3e6f8554594
> 
> I scan the parameter space from 1 to 20 (not even), and the result shows no significant difference.
> 
> So I'm against this kind of modification.

Thanks for your reply,I did have a problem with my test

Reply:
> > @Oliverwqcwrw your test is not measured correctly due to the JVM warmup. The first part will always take much longer unless you run few things before. I also think this load factor should not be tempered here unless there's solid empirical/ mathematical evidence.
> 
> 
> 
> Thanks for your solution, I do wonder why my test result is so strange, it seems that the future test will write a benchmark test

Learn JHM and warnup concept. JVM 's JIT are relative about this

Don't measure one method if it is too fast, try 1000 to 100000 as a unit.

Reply:
Thank you very much. Learned a lot. It's a very friendly community